src/api/box-edit/BoxEdit.js

```diff
@@ -0,0 +1,200 @@
+    const execData = JSON.stringify({
+        auth_code: authCode,
+        auth_token: token,
+        browser_type: Browser.getName(),
+        command_type: 'launch_application',
+        file_id: fileId.toString(),
+        token_scope: tokenScope,
+    });
+    return execData;
+}
+
+function isBlacklistedExtension(extension): boolean {
+    const { EXTENSION_BLACKLIST } = CONSTANTS;
+    let uppercaseExt = extension.toUpperCase();
+
+    // if ext has a leading ., strip it
+    if (uppercaseExt.charAt(0) === '.') {
+        uppercaseExt = uppercaseExt.substr(1);
+    }
+
+    return uppercaseExt in EXTENSION_BLACKLIST;
+}
+
+let BoxEditInstance = null;
+
+class BoxEdit {
+    client: ComServerClient;
+
+    extensionRequestQueue: Map<string, extensionRequestQueueItem>;
+
+    constructor() {
+        if (!(BoxEditInstance instanceof BoxEdit)) {
+            BoxEditInstance = this;
+        }
+        this.extensionRequestQueue = new Map();
+        return BoxEditInstance;
+    }
+
+    queueGetNativeAppNameFromLocal(extension: string): Promise<string> {
+        // There's already a pending or fulfilled request for the appname
+        if (this.extensionRequestQueue.has(extension)) {
+            const queueItem = this.extensionRequestQueue.get(extension);
+            if (!queueItem) {
+                throw new Error('Race condition re: queueGetNativeAppNameFromLocal');
+            }
+            return queueItem.promise;
+        }
+
+        const extensionRequest = {};
+        const appNameRequestPromise = new Promise((resolve, reject) => {
+            extensionRequest.resolve = resolve;
+            extensionRequest.reject = reject;
+        });
+        extensionRequest.promise = appNameRequestPromise;
+        this.extensionRequestQueue.set(extension, extensionRequest);
+
+        return appNameRequestPromise;
+    }
+
+    checkBoxEditAvailability(): Promise<any> {
+        return this.getBoxEditAvailability();
+    }
+
+    getBoxEditAvailability(): Promise<any> {
+        this.client = new ComServerClient(CONSTANTS.BOX_EDIT_APP_NAME);
+
+        return this.client.getComServerStatus();
+    }
+
+    async canOpenWithBoxEdit(extensions: string[]): Promise<any> {
+        const extensionToAppTuples: Array<stringTuple> = await Promise.all(
+            extensions.map(async ext => {
+                try {
+                    const appName = await this.getAppForExtension(ext);
+                    const result: stringTuple = [ext, appName];
+                    return result;
+                } catch (err) {
+                    const result: stringTuple = [ext, ''];
+                    return result;
+                }
+            }),
+        );
+
+        const resultMap: Map<string, string> = new Map();
+        extensionToAppTuples.forEach(tuple => resultMap.set(...tuple));
+
+        return Promise.resolve(resultMap);
+    }
+
+    openFile(fileTypedID: string, token: Object): Promise<any> {
```

Comment:
Flow type for `token`? `tokenDetails` might be a little closer to a correct name, since `token.data.token` sounds strange